README.md
# SuperLogin-client

Javascript bindings to help you build a front-end for SuperLogin applications.
Completely based on its counter part for [AngularJS](https://github.com/colinskow/ng-superlogin) of Colin Skow.

For issues and feature requests visit the [issue tracker](https://github.com/micky2be/superlogin-client/issues).

## Features

- Javascript bindings for all core SuperLogin functionality
- Cookie-less authentication protects your users from CSRF attacks
- Store your authentication token in localStorage or sessionStorage
- HTTP interceptor automatically sends an authentication token to every request to your configured endpoints
- Automatically refreshes the token after a pre-configured portion of its duration has passed
- Emits events to advise your application of any changes to authentication status

## Installation

`npm install superlogin-client --save`

## Configuration

Here is a tour of all the available configuration options:

```js
var config = {
  // An optional URL to API server, by default a current window location is used.
  serverUrl: 'http://localhost:3000',
  // The base URL for the SuperLogin routes with leading and trailing slashes (defaults to '/auth')
  baseUrl: '/auth',
  // Specific endpoint for social authentication and social link popups (defaults to `${location.origin}${baseUrl}`)
  socialUrl: 'http://localhost:3001/auth'
  // A list of API endpoints to automatically add the Authorization header to
  // By default the host the browser is pointed to will be added automatically
  endpoints: ['api.example.com'],
  // Set this to true if you do not want the URL bar host automatically added to the list
  noDefaultEndpoint: false,
  // Where to save your session token: localStorage ('local') or sessionStorage ('session'), default: 'local'
  storage: 'local',
  // The authentication providers that are supported by your SuperLogin host
  providers: ['facebook', 'twitter'],
  // Sets when to check if the session is expired during the setup.
  // false by default.
  checkExpired: false,
  // A float that determines the percentage of a session duration, after which SuperLogin will automatically refresh the
  // token. For example if a token was issued at 1pm and expires at 2pm, and the threshold is 0.5, the token will
  // automatically refresh after 1:30pm. When authenticated, the token expiration is automatically checked on every
  // request. You can do this manually by calling superlogin.checkRefresh(). Default: 0.5
  refreshThreshold: 0.5,
  // The number of milliseconds before a request times out
  // If the request takes longer than `timeout`, the request will be aborted.
  // Default is 0, meaning it won't timeout.
  timeout: 0
};
```
Now let's import SuperLogin and configure it...

```js
import superlogin from 'superlogin-client';
superlogin.configure(config);
```

## Events

##### `login`
```js
superlogin.on('login', function(session) { ... });
```
Session is an object that contains all the session information returned by SuperLogin, along with `serverTimeDiff`, the difference between the server clock and the local clock.

##### `logout`
```js
superlogin.on('logout', function(message) { ... });
```
Message is a message that explains why the user was logged out: 'Logged out' or 'Session expired'.

##### `refresh`
```js
superlogin.on('refresh', function(newSession) { ... });
```
Broadcast when the token is refreshed.

##### `link`
```js
superlogin.on('link', function(provider) { ... });
```
Broadcast when a provider has been linked to the account.

## API

##### `superlogin.authenticated()`
Returns true if the user is currently authenticated; otherwise false. (synchronous)

##### `superlogin.authenticate()`
Returns a promise that is resolved as soon as the user has authenticated.
If the user never authenticates, this promise will stay waiting forever.

##### `superlogin.getConfig()`
Returns the config object. (synchronous)

##### `superlogin.validateSession()`
Makes an HTTP call to verify if the current session is valid

##### `superlogin.getSession()`
Returns the current session if the user is authenticated. (synchronous)

##### `superlogin.deleteSession()`
Deletes the current session, but does not invalidate the token on the server or broadcast a logout event. (synchronous)

##### `superlogin.getDbUrl(dbName)`
Returns the access url for the specified user database, or null if it is not found. (synchronous)

##### `superlogin.confirmRole(role)`
Returns true if the authenticated user possesses the specified `role` (string). (synchronous)

##### `superlogin.confirmAnyRole(possibleRoles)`
Returns true if the user possesses at least one of the specified `possibleRoles` (array). (synchronous)

##### `superlogin.confirmAllRoles(requiredRoles)`
Returns true if the user possesses ALL of the specified `requiredRoles` (array). (synchronous)

##### `superlogin.refresh()`
Makes an HTTP call to refresh the access token.

##### `superlogin.checkRefresh()`
Checks if the session has exceeded the refresh threshold, and calls refresh if necessary

##### `superlogin.checkExpired()`
Checks if the session has expired and logs out if the access token is no longer valid. 
Accounts for server time difference.

##### `superlogin.login(credentials)`
Passes credentials to the server to log the user in.
Returns a promise that resolves with session information upon successful login, or rejects with an error message if login fails.
The credentials object must contain `username` and `password`.

##### `superlogin.register(registration)`
Passes the registration form to SuperLogin to create a new user. Returns a promise.
If the sever returns session information the user is automatically logged in.

##### `superlogin.logout(message)`
Logs out the current session and returns a promise.
Deletes the session and resolves the promise no matter what.
The optional `message` will be broadcast with the 'sl:logout' event.

##### `superlogin.logoutAll(message)`
Logs out ALL the user's open sessions and returns a promise.
Deletes the session and resolves the promise no matter what.
The optional `message` will be broadcast with the 'sl:logout' event.

##### `superlogin.logoutOthers()`
Logs out all the user's open sessions EXCEPT the current one. Returns a promise.

##### `superlogin.socialAuth(provider)`
Opens a popup window to authenticate the specified provider.
Returns a promise that is rejected if authentication fails, or the popup is closed prematurely.
Also rejects if the provider is not present in the `providers` list in the config.

##### `superlogin.tokenSocialAuth(provider, accessToken)`
Login using an access_token obtained by the client for the specified provider.
This is useful for PhoneGap and native plugins.
Rejects if the provider is not present in the `providers` list in the config.

##### `superlogin.link(provider)`
Opens a popup window to link provider to the already authenticated user.
Returns a promise that will reject if the user is not authenticated, the popup is closed prematurely, or the link fails.

##### `superlogin.tokenLink(provider, accessToken)`
Link a provider using an access_token obtained by the client. Returns a promise.

##### `superlogin.unlink(provider)`
Unlinks the specified provider from the user's account. Returns a promise.

##### `superlogin.confirmEmail(token)`
Verifies the user's email with the SuperLogin server, using the specified token.
Returns a promise. Authentication is not required.
The token will be a URL parameter passed in when the user clicks on the confirmation link in the email sent by the system.
Your app needs to manually extract the token from the URL and pass it in here.

##### `superlogin.forgotPassword(email)`
Makes an HTTP request to SuperLogin to send a forgot password email to the user. Authentication is not required.

##### `superlogin.resetPassword(form)`
Forwards the supplied reset password form to SuperLogin. Must include `token`, `password`, and `confirmPassword`.
The token needs to beis extracted from the URL when the user clicks on the link in the password reset email.

##### `superlogin.changePassword(form)`
Changes the authenticated user's password, or creates one if it has not been set.
If a password already exists, then `currentPassword` is required. 
If no password is set, then only `newPassword` and `confirmPassword` are required.

##### `superlogin.changeEmail(newEmail)`
Changes the authenticated user's email.
If email confirmation is enabled, a new confirm email will go out, and the email will not be changed until the new address is confirmed.

##### `superlogin.logoutAll(message)`
Logs out ALL the user's open sessions and returns a promise.
Deletes the session and resolves the promise no matter what.
The optional `message` will be broadcast with the 'sl:logout' event.

##### `superlogin.validateUsername(username)`
Returns a promise that will resolve if the username is valid and not currently in use, or reject otherwise.

##### `superlogin.validateEmail(email)`
Returns a promise that will resolve if the email is valid and not currently in use, or reject otherwise.

##### `superlogin.getHttp()`
Returns the http (Axios) client with the Authorization preset for you.  


CHANGELOG.md
## 0.8.0
- Rightfully changed the API verifyEmail to confirmEnail (by @SteffenDE)

## 0.7.2
- Fixed typescript declaration file with the update of eventemitter2 (by @polco)
- Added support of custom field for authentication (by @ThonyFD)

## 0.7.1
- Added removeItem to the memoryStorage

## 0.7.0
- Updated dependencies

## 0.6.1
- Checking refresh using refreshed date (by @peteruithoven)
- Fixed `socialUrl` option being overwritten (by @atoumbre)

## 0.6.0
- Added Session and ConfigurationOptions types (by @polco)

## 0.5.4
- Added a way to specify a different endpoint for social auth/link

## 0.5.3
- Added a custom timeout option to Axios request (by @peteruithoven)

## 0.5.1-0.5.2
- Using url-parse to parse host for better compatibility (by @peteruithoven)
- The username property is also optional in Typescript definition (by @EuAndreh)
- Added a memoryStorage fallback for private browsing

## 0.5.0
- Dispatch better error when authorization popup is blocked (by @casperlamboo)
- Added serverUrl config option - an optional URL to an external API server (by @mnasyrov)
- Fixed an issue with error object and the status property (by @hacorbin)

## 0.4.3
- Skip refresh in register, forgotPassword and resetPassword (by @peteruithoven)
- Delete session on unauthorized refresh (by @peteruithoven)
- Removed es6-promise from Typescript definition to avoid conflict

## 0.4.2
- Handle undefined error response (by @peteruithoven)
- Linter added
- Refresh session with all new data

## 0.4.1
- Updated typescript definition style

## 0.4.0
- Added support for typescript with definition file

## 0.3.1
- Better error catching (by @peteruithoven)

## 0.3.0
- Fixed debug logger
- Prevent double refresh call (by @peteruithoven)

## 0.2.4
- Return the error object in validateEmail and validatePassword (by @juliobetta)


## 0.2.3
- Avoid 401 error on logout

## 0.2.2
- Code refactoring
- Handle network errors properly (by @Mythli)

## 0.2.1
- Readme update
- Fix issue accessing `this` within setInterval (by @staxmanade)

## 0.2.0
- Added `validateSession` to verify session with the server

## 0.1.X
- Fixed window name for social login
- Removed possible typo reference to Angular (by @SukantGujar)
- Returning session on refresh
- No `login` event on configuration by default
- Using Promise to ensure the API call order
- Reject if there is no session
- Promisify checkRefresh
- Transpile on publish


config.example.js
// Here is a tour of all the available configuration options:
const config = {
	// An optional URL to API server, by default a current window location is used.
	serverUrl: 'http://localhost:3000',
	// The base URL for the SuperLogin routes with leading and trailing slashes (defaults to '/auth/')
	baseUrl: '/auth/',
	// An optional URL for social auth/link, by default equals to `baseUrl`
	socialUrl: 'http://localhost:3001/social',
	// A list of API endpoints to automatically add the Authorization header to
	// By default the host the browser is pointed to will be added automatically
	endpoints: ['api.example.com'],
	// Set this to true if you do not want the URL bar host automatically added to the list
	noDefaultEndpoint: false,
	// Where to save your session token: localStorage ('local') or sessionStorage ('session'), default: 'local'
	storage: 'local',
	// The authentication providers that are supported by your SuperLogin host
	providers: ['facebook', 'twitter'],
	// Sets when to check if the session is expired. 'stateChange', 'startup' or nothing.
	// 'stateChange' checks every time $stateChangeStart or $routeChangeStart is fired
	// 'startup' checks just on app startup. If this is blank it will never check.
	checkExpired: 'stateChange',
	// A float that determines the percentage of a session duration, after which SuperLogin will automatically refresh
	// the token. For example if a token was issued at 1pm and expires at 2pm, and the threshold is 0.5, the token will
	// automatically refresh after 1:30pm. When authenticated, the token expiration is automatically checked on every
	// request. You can do this manually by calling superlogin.checkRefresh(). Default: 0.5
	refreshThreshold: 0.5,
	// The number of milliseconds before a request times out
	// If the request takes longer than `timeout`, the request will be aborted.
	// Default is 0, meaning it won't timeout.
	timeout: 0,
	// For authentication the default fields are `username` and `password`.
	// Use this option to customize it as you did on your server side
	local: {
		usernameField: 'username',
		passwordField: 'password'
	}
};


superlogin-client.d.ts
import { EventEmitter2 } from 'eventemitter2';
import { AxiosInstance } from 'axios';

export type Session = {
	expires: number,
	ip: string,
	password: string,
	provider: string,
	roles: string[],
	serverTimeDiff: number,
	token: string,
	userDBs: { [db: string]: string },
	user_id: string
}

export type ConfigurationOptions = {
	serverUrl?: string,
	baseUrl?: string,
	socialUrl?: string,
	endpoints?: string[],
	noDefaultEndpoint?: boolean,
	storage?: string,
	providers?: string[],
	checkExpired?: boolean,
	refreshThreshold?: number,
	timeout?: number,
	local?: {
		usernameField?: string,
		passwordField?: string
	}
}

export interface SuperLoginClient extends EventEmitter2 {
	configure: (options: ConfigurationOptions) => void;
	authenticated: () => boolean;
	authenticate: () => Promise<any>;
	getConfig: () => ConfigurationOptions;
	validateSession: () => void;
	getSession: () => Session;
	setSession: (session: Session) => void;
	deleteSession: () => void;
	getDbUrl: (dbName: string) => string;
	confirmRole: (role: string) => boolean;
	confirmAnyRole: (possibleRoles: string[]) => boolean;
	confirmAllRoles: (requiredRoles: string[]) => boolean;
	refresh: () => void;
	checkRefresh: () => void;
	checkExpired: () => void;
	login: (login: { username: string, password: string }) => Promise<any>;
	register: (register: { username?: string, name?: string, email?: string, password: string, confirmPassword: string, [key: string]: any }) => Promise<any>;
	logout: (message?: string) => Promise<any>;
	logoutAll: (message?: string) => Promise<any>;
	logoutOthers: () => Promise<any>;
	socialAuth: (provider: string) => Promise<any>;
	tokenSocialAuth: (provider: string, accessToken: string) => Promise<any>;
	link: (provider: string) => Promise<any>;
	tokenLink: (provider: string, accessToken: string) => Promise<any>;
	unlink: (provider: string) => Promise<any>;
	verifyEmail: (token: string) => Promise<any>;
	forgotPassword: (email: string) => Promise<any>;
	resetPassword: (reset: { password: string, confirmPassword: string, token: string }) => Promise<any>;
	changePassword: (change: { currentPassword?: string, newPassword: string, confirmPassword: string, token: string }) => Promise<any>;
	changeEmail: (email: string) => Promise<any>;
	validateUsername: (username: string) => Promise<any>;
	validateEmail: (email: string) => Promise<any>;
	getHttp: () => AxiosInstance;
}

export as namespace SuperLogin;
declare const client: SuperLoginClient;
export default client;


src/index.js
import axios from 'axios';
import _debug from 'debug';
import { EventEmitter2 } from 'eventemitter2';
import URL from 'url-parse';

const debug = {
	log: _debug('superlogin:log'),
	info: _debug('superlogin:info'),
	warn: _debug('superlogin:warn'),
	error: _debug('superlogin:error')
};

// Capitalizes the first letter of a string
function capitalizeFirstLetter(string) {
	return string.charAt(0).toUpperCase() + string.slice(1);
}

function parseHostFromUrl(url) {
	const parsedURL = new URL(url);
	return parsedURL.host;
}

function checkEndpoint(url, endpoints) {
	const host = parseHostFromUrl(url);
	for (let i = 0; i < endpoints.length; i += 1) {
		if (host === endpoints[i]) {
			return true;
		}
	}
	return false;
}
function isStorageAvailable() {
	const mod = '__STORAGE__';
	try {
		localStorage.setItem(mod, mod);
		localStorage.removeItem(mod);
		return true;
	} catch (e) {
		return false;
	}
}

function parseError(err) {
	// if no connection can be established we don't have any data thus we need to forward the original error.
	if (err && err.response && err.response.data) {
		return err.response.data;
	}
	return err;
}

const memoryStorage = {
	setItem: (key, value) => {
		memoryStorage.storage.set(key, value);
	},
	getItem: key => {
		const value = memoryStorage.storage.get(key);
		if (typeof value !== 'undefined') {
			return value;
		}
		return null;
	},
	removeItem: key => {
		memoryStorage.storage.delete(key);
	},
	storage: new Map()
};

class Superlogin extends EventEmitter2 {
	constructor() {
		super();

		this._oauthComplete = false;
		this._config = {};
		this._refreshInProgress = false;
		this._http = axios.create();
	}

	configure(config = {}) {
		if (config.serverUrl) {
			this._http = axios.create({
				baseURL: config.serverUrl,
				timeout: config.timeout
			});
		}

		config.baseUrl = config.baseUrl || '/auth';
		config.baseUrl = config.baseUrl.replace(/\/$/, ''); // remove trailing /
		config.socialUrl = config.socialUrl || config.baseUrl;
		config.socialUrl = config.socialUrl.replace(/\/$/, ''); // remove trailing /
		config.local = config.local || {};
		config.local.usernameField = config.local.usernameField || 'username';
		config.local.passwordField = config.local.passwordField || 'password';

		if (!config.endpoints || !(config.endpoints instanceof Array)) {
			config.endpoints = [];
		}
		if (!config.noDefaultEndpoint) {
			let defaultEndpoint = window.location.host;
			if (config.serverUrl) {
				defaultEndpoint = parseHostFromUrl(config.serverUrl);
			}
			config.endpoints.push(defaultEndpoint);
		}
		config.providers = config.providers || [];
		config.timeout = config.timeout || 0;

		if (!isStorageAvailable()) {
			this.storage = memoryStorage;
		} else if (config.storage === 'session') {
			this.storage = window.sessionStorage;
		} else {
			this.storage = window.localStorage;
		}

		this._config = config;

		// Setup the new session
		this._session = JSON.parse(this.storage.getItem('superlogin.session'));

		this._httpInterceptor();

		// Check expired
		if (config.checkExpired) {
			this.checkExpired();
			this.validateSession()
				.then(() => {
					this._onLogin(this._session);
				})
				.catch(() => {
					// ignoring
				});
		}
	}

	_httpInterceptor() {
		const request = req => {
			const config = this.getConfig();
			const session = this.getSession();
			if (!session || !session.token) {
				return Promise.resolve(req);
			}

			if (req.skipRefresh) {
				return Promise.resolve(req);
			}

			return this.checkRefresh().then(() => {
				if (checkEndpoint(req.url, config.endpoints)) {
					req.headers.Authorization = `Bearer ${session.token}:${session.password}`;
				}
				return req;
			});
		};

		const responseError = error => {
			const config = this.getConfig();

			// if there is not config obj in in the error it means we cannot check the endpoints.
			// This happens for example if there is no connection at all because axion just forwards the raw error.
			if (!error || !error.config) {
				return Promise.reject(error);
			}

			// If there is an unauthorized error from one of our endpoints and we are logged in...
			if (checkEndpoint(error.config.url, config.endpoints) &&
				error.response && error.response.status === 401 && this.authenticated()) {
				debug.warn('Not authorized');
				this._onLogout('Session expired');
			}
			return Promise.reject(error);
		};
		// clear interceptors from a previous configure call
		this._http.interceptors.request.eject(this._httpRequestInterceptor);
		this._http.interceptors.response.eject(this._httpResponseInterceptor);

		this._httpRequestInterceptor = this._http.interceptors.request.use(request.bind(this));
		this._httpResponseInterceptor = this._http.interceptors.response.use(null, responseError.bind(this));
	}

	authenticated() {
		return !!(this._session && this._session.user_id);
	}

	getConfig() {
		return this._config;
	}

	validateSession() {
		if (!this.authenticated()) {
			return Promise.reject();
		}
		return this._http.get(`${this._config.baseUrl}/session`)
			.catch(err => {
				this._onLogout('Session expired');
				throw parseError(err);
			});
	}

	getSession() {
		if (!this._session) {
			this._session = JSON.parse(this.storage.getItem('superlogin.session'));
		}
		return this._session ? Object.assign(this._session) : null;
	}

	setSession(session) {
		this._session = session;
		this.storage.setItem('superlogin.session', JSON.stringify(this._session));
		debug.info('New session set');
	}

	deleteSession() {
		this.storage.removeItem('superlogin.session');
		this._session = null;
	}

	getDbUrl(dbName) {
		if (this._session.userDBs && this._session.userDBs[dbName]) {
			return this._session.userDBs[dbName];
		}
		return null;
	}

	getHttp() {
		return this._http;
	}

	confirmRole(role) {
		if (!this._session || !this._session.roles || !this._session.roles.length) return false;
		return this._session.roles.indexOf(role) !== -1;
	}

	confirmAnyRole(roles) {
		if (!this._session || !this._session.roles || !this._session.roles.length) return false;
		for (let i = 0; i < roles.length; i += 1) {
			if (this._session.roles.indexOf(roles[i]) !== -1) return true;
		}
		return false;
	}

	confirmAllRoles(roles) {
		if (!this._session || !this._session.roles || !this._session.roles.length) return false;
		for (let i = 0; i < roles.length; i += 1) {
			if (this._session.roles.indexOf(roles[i]) === -1) return false;
		}
		return true;
	}

	checkRefresh() {
		// Get out if we are not authenticated or a refresh is already in progress
		if (this._refreshInProgress) {
			return Promise.resolve();
		}
		if (!this._session || !this._session.user_id) {
			return Promise.reject();
		}
		// try getting the latest refresh date, if not available fall back to issued date
		const refreshed = this._session.refreshed || this._session.issued;
		const expires = this._session.expires;
		const threshold = isNaN(this._config.refreshThreshold) ? 0.5 : this._config.refreshThreshold;
		const duration = expires - refreshed;
		let timeDiff = this._session.serverTimeDiff || 0;
		if (Math.abs(timeDiff) < 5000) {
			timeDiff = 0;
		}
		const estimatedServerTime = Date.now() + timeDiff;
		const elapsed = estimatedServerTime - refreshed;
		const ratio = elapsed / duration;
		if (ratio > threshold) {
			debug.info('Refreshing session');
			return this.refresh()
				.then(session => {
					debug.log('Refreshing session sucess', session);
					return session;
				})
				.catch(err => {
					debug.error('Refreshing session failed', err);
					throw err;
				});
		}
		return Promise.resolve();
	}

	checkExpired() {
		// This is not necessary if we are not authenticated
		if (!this.authenticated()) {
			return;
		}
		const expires = this._session.expires;
		let timeDiff = this._session.serverTimeDiff || 0;
		// Only compensate for time difference if it is greater than 5 seconds
		if (Math.abs(timeDiff) < 5000) {
			timeDiff = 0;
		}
		const estimatedServerTime = Date.now() + timeDiff;
		if (estimatedServerTime > expires) {
			this._onLogout('Session expired');
		}
	}

	refresh() {
		const session = this.getSession();
		this._refreshInProgress = true;
		return this._http.post(`${this._config.baseUrl}/refresh`, {})
			.then(res => {
				this._refreshInProgress = false;
				if (res.data.token && res.data.expires) {
					Object.assign(session, res.data);
					this.setSession(session);
					this._onRefresh(session);
				}
				return session;
			})
			.catch(err => {
				this._refreshInProgress = false;
				throw parseError(err);
			});
	}

	authenticate() {
		return new Promise(resolve => {
			const session = this.getSession();
			if (session) {
				resolve(session);
			} else {
				this.on('login', newSession => {
					resolve(newSession);
				});
			}
		});
	}

	login(credentials) {
		const { usernameField, passwordField } = this._config.local;
		if (!credentials[usernameField] || !credentials[passwordField]) {
			return Promise.reject({ error: 'Username or Password missing...' });
		}
		return this._http.post(`${this._config.baseUrl}/login`, credentials, { skipRefresh: true })
			.then(res => {
				res.data.serverTimeDiff = res.data.issued - Date.now();
				this.setSession(res.data);
				this._onLogin(res.data);
				return res.data;
			})
			.catch(err => {
				this.deleteSession();

				throw parseError(err);
			});
	}

	register(registration) {
		return this._http.post(`${this._config.baseUrl}/register`, registration, { skipRefresh: true })
			.then(res => {
				if (res.data.user_id && res.data.token) {
					res.data.serverTimeDiff = res.data.issued - Date.now();
					this.setSession(res.data);
					this._onLogin(res.data);
				}
				this._onRegister(registration);
				return res.data;
			})
			.catch(err => {
				throw parseError(err);
			});
	}

	logout(msg) {
		return this._http.post(`${this._config.baseUrl}/logout`, {})
			.then(res => {
				this._onLogout(msg || 'Logged out');
				return res.data;
			})
			.catch(err => {
				this._onLogout(msg || 'Logged out');
				if (!err.response || err.response.data.status !== 401) {
					throw parseError(err);
				}
			});
	}

	logoutAll(msg) {
		return this._http.post(`${this._config.baseUrl}/logout-all`, {})
			.then(res => {
				this._onLogout(msg || 'Logged out');
				return res.data;
			})
			.catch(err => {
				this._onLogout(msg || 'Logged out');
				if (!err.response || err.response.data.status !== 401) {
					throw parseError(err);
				}
			});
	}

	logoutOthers() {
		return this._http.post(`${this._config.baseUrl}/logout-others`, {})
			.then(res => res.data)
			.catch(err => {
				throw parseError(err);
			});
	}

	socialAuth(provider) {
		const providers = this._config.providers;
		if (providers.indexOf(provider) === -1) {
			return Promise.reject({ error: `Provider ${provider} not supported.` });
		}
		const url = `${this._config.socialUrl}/${provider}`;
		return this._oAuthPopup(url, { windowTitle: `Login with ${capitalizeFirstLetter(provider)}` });
	}

	tokenSocialAuth(provider, accessToken) {
		const providers = this._config.providers;
		if (providers.indexOf(provider) === -1) {
			return Promise.reject({ error: `Provider ${provider} not supported.` });
		}
		return this._http.post(`${this._config.baseUrl}/${provider}/token`, { access_token: accessToken })
			.then(res => {
				if (res.data.user_id && res.data.token) {
					res.data.serverTimeDiff = res.data.issued - Date.now();
					this.setSession(res.data);
					this._onLogin(res.data);
				}
				return res.data;
			})
			.catch(err => {
				throw parseError(err);
			});
	}

	tokenLink(provider, accessToken) {
		const providers = this._config.providers;
		if (providers.indexOf(provider) === -1) {
			return Promise.reject({ error: `Provider ${provider} not supported.` });
		}
		const linkURL = `${this._config.baseUrl}/link/${provider}/token`;
		return this._http.post(linkURL, { access_token: accessToken })
			.then(res => res.data)
			.catch(err => {
				throw parseError(err);
			});
	}

	link(provider) {
		const providers = this._config.providers;
		if (providers.indexOf(provider) === -1) {
			return Promise.reject({ error: `Provider ${provider} not supported.` });
		}
		if (this.authenticated()) {
			const session = this.getSession();
			const token = `bearer_token=${session.token}:${session.password}`;
			const linkURL = `${this._config.socialUrl}/link/${provider}?${token}`;
			const windowTitle = `Link your account to ${capitalizeFirstLetter(provider)}`;
			return this._oAuthPopup(linkURL, { windowTitle });
		}
		return Promise.reject({ error: 'Authentication required' });
	}

	unlink(provider) {
		const providers = this._config.providers;
		if (providers.indexOf(provider) === -1) {
			return Promise.reject({ error: `Provider ${provider} not supported.` });
		}
		if (this.authenticated()) {
			return this._http.post(`${this._config.baseUrl}/unlink/${provider}`)
				.then(res => res.data)
				.catch(err => {
					throw parseError(err);
				});
		}
		return Promise.reject({ error: 'Authentication required' });
	}

	confirmEmail(token) {
		if (!token || typeof token !== 'string') {
			return Promise.reject({ error: 'Invalid token' });
		}
		return this._http.get(`${this._config.baseUrl}/confirm-email/${token}`)
			.then(res => res.data)
			.catch(err => {
				throw parseError(err);
			});
	}

	forgotPassword(email) {
		return this._http.post(`${this._config.baseUrl}/forgot-password`, { email }, { skipRefresh: true })
			.then(res => res.data)
			.catch(err => {
				throw parseError(err);
			});
	}

	resetPassword(form) {
		return this._http.post(`${this._config.baseUrl}/password-reset`, form, { skipRefresh: true })
			.then(res => {
				if (res.data.user_id && res.data.token) {
					this.setSession(res.data);
					this._onLogin(res.data);
				}
				return res.data;
			})
			.catch(err => {
				throw parseError(err);
			});
	}

	changePassword(form) {
		if (this.authenticated()) {
			return this._http.post(`${this._config.baseUrl}/password-change`, form)
				.then(res => res.data)
				.catch(err => {
					throw parseError(err);
				});
		}
		return Promise.reject({ error: 'Authentication required' });
	}

	changeEmail(newEmail) {
		if (this.authenticated()) {
			return this._http.post(`${this._config.baseUrl}/change-email`, { newEmail })
				.then(res => res.data)
				.catch(err => {
					throw parseError(err);
				});
		}
		return Promise.reject({ error: 'Authentication required' });
	}

	validateUsername(username) {
		return this._http.get(`${this._config.baseUrl}/validate-username/${encodeURIComponent(username)}`)
			.then(() => true)
			.catch(err => {
				throw parseError(err);
			});
	}

	validateEmail(email) {
		return this._http.get(`${this._config.baseUrl}/validate-email/${encodeURIComponent(email)}`)
			.then(() => true)
			.catch(err => {
				throw parseError(err);
			});
	}

	_oAuthPopup(url, options) {
		return new Promise((resolve, reject) => {
			this._oauthComplete = false;
			options.windowName = options.windowTitle ||	'Social Login';
			options.windowOptions = options.windowOptions || 'location=0,status=0,width=800,height=600';
			const _oauthWindow = window.open(url, options.windowName, options.windowOptions);

			if (!_oauthWindow) {
				reject({ error: 'Authorization popup blocked' });
			}

			const _oauthInterval = setInterval(() => {
				if (_oauthWindow.closed) {
					clearInterval(_oauthInterval);
					if (!this._oauthComplete) {
						this.authComplete = true;
						reject({ error: 'Authorization cancelled' });
					}
				}
			}, 500);

			window.superlogin = {};
			window.superlogin.oauthSession = (error, session, link) => {
				if (!error && session) {
					session.serverTimeDiff = session.issued - Date.now();
					this.setSession(session);
					this._onLogin(session);
					return resolve(session);
				} else if (!error && link) {
					this._onLink(link);
					return resolve(`${capitalizeFirstLetter(link)} successfully linked.`);
				}
				this._oauthComplete = true;
				return reject(error);
			};
		});
	}

	_onLogin(msg) {
		debug.info('Login', msg);
		this.emit('login', msg);
	}

	_onLogout(msg) {
		this.deleteSession();
		debug.info('Logout', msg);
		this.emit('logout', msg);
	}

	_onLink(msg) {
		debug.info('Link', msg);
		this.emit('link', msg);
	}

	_onRegister(msg) {
		debug.info('Register', msg);
		this.emit('register', msg);
	}

	_onRefresh(msg) {
		debug.info('Refresh', msg);
		this.emit('refresh', msg);
	}
}


export default new Superlogin();
